File: apps/api/src/main/java/com/demo/insight/user/service/UserService.java
package com.demo.insight.user.service;

import com.demo.insight.auth.service.RequestActorService;
import com.demo.insight.user.dto.ChangePasswordRequestDto;
import com.demo.insight.user.dto.CreateUserRequestDto;
import com.demo.insight.user.dto.UpdateUserRequestDto;
import com.demo.insight.user.dto.UserDto;

import java.util.List;

public interface UserService {

    List<UserDto> getUsers(RequestActorService.AuthenticatedActor actor);

    UserDto getUser(String empcode, RequestActorService.AuthenticatedActor actor);

    UserDto createUser(CreateUserRequestDto request, RequestActorService.AuthenticatedActor actor);

    UserDto updateUser(String empcode, UpdateUserRequestDto request, RequestActorService.AuthenticatedActor actor);

    void deleteUser(String empcode, RequestActorService.AuthenticatedActor actor);

    void changePassword(String empcode, ChangePasswordRequestDto request, RequestActorService.AuthenticatedActor actor);
}

File: apps/web/src/services/userService.ts
import { apiClient } from './apiClient';
import type {
  ChangePasswordRequest,
  CreateUserRequest,
  UpdateUserRequest,
  UserItem,
} from '../types/user';

export const userService = {
  getUsers() {
    return apiClient.request<UserItem[]>('/api/users');
  },

  getUser(empcode: string) {
    const encodedEmpcode = encodeURIComponent(empcode);
    return apiClient.request<UserItem>(`/api/users/${encodedEmpcode}`);
  },

  createUser(payload: CreateUserRequest) {
    return apiClient.request<UserItem>('/api/users', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  updateUser(empcode: string, payload: UpdateUserRequest) {
    const encodedEmpcode = encodeURIComponent(empcode);
    return apiClient.request<UserItem>(`/api/users/${encodedEmpcode}`, {
      method: 'PUT',
      body: JSON.stringify(payload),
    });
  },

  deleteUser(empcode: string) {
    const encodedEmpcode = encodeURIComponent(empcode);
    return apiClient.request<void>(`/api/users/${encodedEmpcode}`, {
      method: 'DELETE',
    });
  },

  changePassword(empcode: string, payload: ChangePasswordRequest) {
    const encodedEmpcode = encodeURIComponent(empcode);
    return apiClient.request<void>(`/api/users/${encodedEmpcode}/password`, {
      method: 'PUT',
      body: JSON.stringify(payload),
    });
  },
};
